experimental/features.py

```diff
@@ -0,0 +1,371 @@
+                            kappa0_feat).reshape(input_shape + (-1,))
+
+    elif method == 'ps':
+      # ps0: PolyTensorSRHT = input[0]
+      # ps1: PolyTensorSRHT = input[1]
+      # ts2: TensorSRHT2 = input[2]
+      raise NotImplementedError
+
+    elif method == 'exact':  # Exact feature extraction via Cholesky decomposition.
+      nngp_feat = cholesky(kappa1(nngp_feat_2d)).reshape(input_shape + (-1,))
+
+      ntk = ntk_feat_2d @ ntk_feat_2d.T
+      kappa0_mat = kappa0(nngp_feat_2d)
+      ntk_feat = cholesky(ntk * kappa0_mat).reshape(input_shape + (-1,))
+
+    else:
+      raise NotImplementedError
+
+    return f.replace(nngp_feat=nngp_feat, ntk_feat=ntk_feat)
+
+  return init_fn, feature_fn
+
+
+def _conv_feat(X, filter_size):
+  N, H, W, C = X.shape
+  out = np.zeros((N, H, W, C * filter_size))
+  out = out.at[:, :, :, :C].set(X)
+  j = 1
+  for i in range(1, min((filter_size + 1) // 2, W)):
+    out = out.at[:, :, :-i, j * C:(j + 1) * C].set(X[:, :, i:])
+    j += 1
+    out = out.at[:, :, i:, j * C:(j + 1) * C].set(X[:, :, :-i])
+    j += 1
+  return out
+
+
+def _conv2d_feat(X, filter_size):
+  return _conv_feat(np.moveaxis(_conv_feat(X, filter_size), 1, 2), filter_size)
+
+
+@layer
+def ConvFeatures(out_dim: int,
+                 filter_size: int,
+                 W_std: float = 1.0,
+                 b_std: float = 0.,
+                 channel_axis: int = -1):
+
+  if b_std != 0.0:
+    raise NotImplementedError('Non-zero b_std is not implemented yet .'
+                              ' Please set b_std to be `0`.')
+
+  def init_fn(rng, input_shape):
+    nngp_feat_shape, ntk_feat_shape = input_shape[0], input_shape[1]
+    new_nngp_feat_shape = nngp_feat_shape[:-1] + (nngp_feat_shape[-1] *
+                                                  filter_size**2,)
+    new_ntk_feat_shape = nngp_feat_shape[:-1] + (
+        (nngp_feat_shape[-1] + ntk_feat_shape[-1]) * filter_size**2,)
+    return (new_nngp_feat_shape, new_ntk_feat_shape), ()
+
+  def feature_fn(f, input, **kwargs):
+    nngp_feat, ntk_feat = f.nngp_feat, f.ntk_feat
+
+    nngp_feat = _conv2d_feat(nngp_feat, filter_size) / filter_size * W_std
+
+    if _is_defaut_feature(ntk_feat):  # check if ntk_feat is empty
+      ntk_feat = nngp_feat
+    else:
+      ntk_feat = _conv2d_feat(ntk_feat, filter_size) / filter_size * W_std
+      ntk_feat = np.concatenate((ntk_feat, nngp_feat), axis=channel_axis)
+
+    return f.replace(nngp_feat=nngp_feat, ntk_feat=ntk_feat)
+
+  return init_fn, feature_fn
+
+
+@layer
+def AvgPoolFeatures(window_size: int,
+                    stride_size: int = 2,
+                    padding: str = stax.Padding.VALID.name,
+                    normalize_edges: bool = False,
+                    batch_axis: int = 0,
+                    channel_axis: int = -1):
+
+  def init_fn(rng, input_shape):
+    nngp_feat_shape, ntk_feat_shape = input_shape[0], input_shape[1]
+
+    new_nngp_feat_shape = nngp_feat_shape[:1] + (
```

Comment:
Here and in the CNN case, to better support various padding/strides/dilation/etc, you may want to use `jax.eval_shape` to compute the output shape (see example https://github.com/google/neural-tangents/blob/dcf4bc96e72761c9d350b79ba93c976ab2f9f1c8/neural_tangents/_src/stax/linear.py#L1107). Feel free to leave for later.